src/crimson/os/seastore/btree/fixed_kv_btree.h

```diff
@@ -1749,124 +1933,145 @@ class FixedKVBtree {
+        auto [left, right, pivot] = pos.node->make_split_children(c);
 
-    for (; split_from > 0; --split_from) {
-      auto &parent_pos = iter.get_internal(split_from + 1);
-      if (!parent_pos.node->is_mutable()) {
-        parent_pos.node = c.cache.duplicate_for_write(
-          c.trans, parent_pos.node
-        )->template cast<internal_node_t>();
-      }
+        auto parent_node = parent_pos.node;
+        auto parent_iter = parent_pos.get_iter();
+
+        parent_node->update(
+          parent_iter,
+          left->get_paddr(),
+          left.get());
+        parent_node->insert(
+          parent_iter + 1,
+          pivot,
+          right->get_paddr(),
+          right.get());
 
-      if (split_from > 1) {
-        auto &pos = iter.get_internal(split_from);
         SUBTRACET(
           seastore_fixedkv_tree,
-          "splitting internal {} at depth {}, parent: {} at pos: {}",
+          "splitted {} into left: {}, right: {}",
           c.trans,
           *pos.node,
-          split_from,
-          *parent_pos.node,
-          parent_pos.pos);
-        auto [left, right] = split_level(parent_pos, pos);
+          *left,
+          *right);
+        c.cache.retire_extent(c.trans, pos.node);
 
-        if (pos.pos < left->get_size()) {
-          pos.node = left;
-        } else {
-          pos.node = right;
-          pos.pos -= left->get_size();
+        get_tree_stats<self_type>(c.trans).extents_num_delta++;
+        return std::make_pair(left, right);
+      };
 
-          parent_pos.pos += 1;
-        }
-      } else {
-        auto &pos = iter.leaf;
-        SUBTRACET(
-          seastore_fixedkv_tree,
-          "splitting leaf {}, parent: {} at pos: {}",
-          c.trans,
-          *pos.node,
-          *parent_pos.node,
-          parent_pos.pos);
-        auto [left, right] = split_level(parent_pos, pos);
-
-        /* right->get_node_meta().begin == pivot == right->begin()->get_key()
-         * Thus, if pos.pos == left->get_size(), we want iter to point to
-         * left with pos.pos at the end rather than right with pos.pos = 0
-         * since the insertion would be to the left of the first element
-         * of right and thus necessarily less than right->get_node_meta().begin.
-         */
-        if (pos.pos <= left->get_size()) {
-          pos.node = left;
-        } else {
-          pos.node = right;
-          pos.pos -= left->get_size();
+      return seastar::do_with(
+        split_from,
+        std::move(split_level),
+        [FNAME, &iter, c](auto &split_from, auto &split_level) {
+        return trans_intr::repeat(
+          [&split_from, &iter, c, FNAME, &split_level] {
+          if (split_from == 0) {
+            return iterator::ensure_internal_iertr::template make_ready_future<
+              seastar::stop_iteration>(seastar::stop_iteration::yes);
+          }
+          return iter.ensure_internal(c, split_from + 1
```

Comment:
Roughly looked at the diff here, can we make sure `ensure_internal()` is called at the beginning with `check_split()`, so that the following logic can be mostly unchanged and keeps to be synchronous (which is a good property), because this is top-down ?

If the above is possible, we can split out 1 cleanup commits to be helpful:
1. a cleanup-only preparation: make `check_split()` interface asynchronous, and only moving the code.
2. apply real changes after, with smaller diff to look at.